API: restructure and document numpy.random C-API

mattip opened the pull request:
Finish the C-API side of the new numpy.random API

mattip committed:
API: restructure and document numpy.random C-API

mattip commented:
Still a WIP. Need to decide how much of the `*.pxd/*.h/*.so` to expose (and commit to supporting, since exposing it means it is now publc). For instance, we currently do not build a `distributions` pxd file, and it seems the `src/aligned_malloc` files do not really belong to random.

mattip commented:
xref gh-5312 for moving the aligned allocations into numpy itself

mattip commented:
Sorry this is a bit long for a PR comment. Should it be an issue?

[Line 75](https://github.com/numpy/numpy/pull/14604/files#diff-c4427b946302b09cddcda541a9e9dfdaR75) says "An example ... is in the examples folder" which is part of the sources under `numpy.random` but not distributed in the wheel.

It contains examples of using the BitGenerator interfaces `next_double` and `next_uint32` and the Generator from cython. To support this via `cimport`, we should ship a `_bit_generator.pxd` (which we already do) and a `_generator.pxd` (TBD). Can we remove the use of PyCapsule since we can directly cimport the BitGenerators?

I am not a numba expert. One of the examples is very close to the one in the `exending.rst` document, so I think is covered. [The other in that directory](https://github.com/numpy/numpy/blob/master/numpy/random/examples/numba/extending_distributions.py) seems to be a cffi example, not a numba one. It is simple enough we could put it in the `extending.rst` document.

TBD:
- [ ] add a `_generator.pxd` file for cython
  - with or without an accompanying `distributions.h` file? If *with* then where should it live?
- [ ] add a cffi example to `extending.rst`
- [ ] change the examples to not use PyCapsule (is this possible and advisable?)

Are we answering the user stories in these examples? 

One of the mailing list comments linked to the [scipy.special](http://scipy.github.io/devdocs/special.cython_special.html) documentation, should we try to create a page like that?

#### User stories - we should try to relate to these (from the [mailing list thread](http://numpy-discussion.10968.n7.nabble.com/Low-level-API-for-Random-td47560.html))

Someone asked how to use random in a ufunc.

"An ideal API would allow projects like https://github.com/deepmind/torch-randomkit/tree/master/randomkit or numba to consume the code in NumPy without vendoring it."

"There are c++ applications which use boost::random, would be nice to be able to swap it for numpy.random."

"Using the existing distributions from Cython was a requested feature and an explicit goal, yes. There are users waiting for this." (mattip: but isn't this supported via `Generator`?)

"Numba would definitely appreciate C functions to access the random distribution implementations, and have a side-project (numba-scipy) that is making the Cython wrapped functions in SciPy visible to Numba" (mattip: I think we do this after gh-14608 from _generator.so via these [`cdef extern` declarations](https://github.com/numpy/numpy/blob/master/numpy/random/_generator.pyx#L54), but that requires the H file.

Maybe someone who wants to write a new BitGenerator, i.e., the [numpy/bitgenerator repo](https://github.com/numpy/bitgenerators) without needing to have the numpy code as a git submodule

rgommers commented on doc/source/reference/random/extending.rst:
```diff
@@ -99,7 +111,23 @@ removing bounds checks and wrap around, providing array alignment information
+
+    def normals_zig(Py_ssize_t n):
+
                 random_values[i] = random_gauss_zig(rng)
```
the `_zig` is no longer in the name

mattip replied:
this whole block should be included via a `.. literalinclude` directive from the appropriate example file

rgommers commented:
> Sorry this is a bit long for a PR comment. Should it be an issue?

Might indeed make sense to merge this quickly (there's only some doc fixes) here that improve the current state of `master`), and then use a new issue for the rest.

> An example ... is in the examples folder

Good point. I'd suggest moving the examples into the documentation, shipping them in the tarball and including them in the rendered docs as well.

> One of the mailing list comments linked to the [scipy.special](http://scipy.github.io/devdocs/special.cython_special.html) documentation, should we try to create a page like that?

I think so. It's no different than any other public function in the Python and C API: every function and object should be listed.

> User stories - we should try to relate to these

Nice collection. Yes, those would make good examples/subsections in the user guide section on this topic.

mattip committed:
DOC: fix bad reference

mattip committed:
API: ship, document, and start to test numpy.random C-API examples

mattip commented on doc/source/reference/random/bit_generators/index.rst:
```diff
@@ -19,7 +19,7 @@ The included BitGenerators are:
   :meth:`~.PCG64.advance`. PCG-64 has a period of :math:`2^{128}`. See the `PCG
   author's page`_ for more details about this class of PRNG.
-* MT19937 - The standard Python BitGenerator. Adds a `~mt19937.MT19937.jumped`
+* MT19937 - The standard Python BitGenerator. Adds a `MT19937.jumped`
```
after gh-14699, I could see this bad ref that I missed in gh-14608. Not really part of this PR, but doesn't justify its own PR.

mattip commented on doc/source/reference/random/index.rst:
```diff
@@ -32,7 +32,7 @@ instance's methods are imported into the numpy.random namespace, see
 -----------
 
-By default, `~Generator` uses bits provided by `~pcg64.PCG64` which
+By default, `~Generator` uses bits provided by `PCG64` which
```
see comment elsewhere about bad refs from #14608

mattip commented:
The cython code does not compile, it needs `include/bitgen.h`. Now we need to decide how (and if) to add these include files, or if there is a way to work around it

mattip committed:
API, DOC, TST: fix tests, refactor documentation to include snippets

mattip committed:
BUILD: move public headers to numpy/core/include/numpy/random

mattip committed:
TST: ignore DeprecationWarnings in setuptools and numba

mattip commented:
Something is fishy with macos Azure runs. It is printing:
```
Installed /Users/runner/hostedtoolcache/Python/3.6.9/x64/lib/python3.6/site-packages/numpy-1.18.0.dev0+20c8d7c-py3.6-macosx-10.14-x86_64.egg
Processing dependencies for numpy==1.18.0.dev0+20c8d7c
Searching for numpy==1.18.0.dev0+20c8d7c
Reading https://pypi.org/simple/numpy/
No local packages or working download links found for numpy==1.18.0.dev0+20c8d7c
Running from numpy source directory.
/Users/runner/hostedtoolcache/Python/3.6.9/x64/lib/python3.6/distutils/dist.py:261: UserWarning: Unknown distribution option: 'define_macros'
  warnings.warn(msg)
error: Could not find suitable distribution for Requirement.parse('numpy==1.18.0.dev0+20c8d7c')
```
I think this is the result of `python setup.py build -j 4 build_src --verbose-cfg install`, but when I run that locally on python3.6, ubuntu 18.04 it runs to completion.

Anyone with MacOS care to try to duplicate it? @tylerjereddy any ideas?

mattip commented:
In 37bbe58 I added `include/bitgen.h` and `include/distributions.h`. Thoughts?

mattip committed:
DOC: document the C-API as used from Cython

mattip commented:
rebased off master to resolve conflicts after merge of #13794. Also:
- moved public random headers to `numpy/core/include/numpy/ranom`
- add first draft of random C-API public methods, rendered [here](https://9926-908607-gh.circle-artifacts.com/0/home/circleci/repo/doc/build/html/reference/random/c-api.html)

mattip commented:
I removed the WIP from the title. The example works, we have the new c-api documentation (link in the comment above) and examples both in [documentation](https://9926-908607-gh.circle-artifacts.com/0/home/circleci/repo/doc/build/html/reference/random/extending.html) and in tests. Reviews would be welcome. More examples and documentation will be generated in gh-14778

charris commented on doc/source/reference/random/c-api.rst:
```diff
@@ -0,0 +1,177 @@
+
+.. currentmodule:: numpy.random
+
+Typed versions of many of the `Generator` and `BitGenerator` can be accessed
```
This is confusing. What does it mean?

mattip replied:
missing fragment. It should read "Typed versions of many of the methods and functions in `Generator` and `BitGenerator` can be accessed ...
Fixing

charris commented on doc/source/reference/random/c-api.rst:
```diff
@@ -0,0 +1,177 @@
+Typed versions of many of the `Generator` and `BitGenerator` can be accessed
+directly from Cython: the complete list is given below.
+
+The ``_bit_generator`` module is usable via::
```
Maybe "is available".

mattip replied:
fixed, and again below

mattip commented:
Rebased to clear a merge confilict. @rkern @bashtage could you review this?

charris commented:
Thanks Matti. I'm not convinced that the examples are in the right place but they can probably be moved later without problems.

charris commented:
The examples might better be included somewhere in the `/doc` directory, the current location might be a development artifact. @rgommers Any thoughts on the examples?

mattip commented:
I put the examples there so they could be tested. Pieces of them are `included` in the docs

rgommers commented:
> I put the examples there so they could be tested. Pieces of them are included in the docs

That sounds like a good idea actually. Please move them to `_examples` though. There's a reason I added a test for unwanted public submodule API changes recently, you should never have to add to the whitelist in that test.
